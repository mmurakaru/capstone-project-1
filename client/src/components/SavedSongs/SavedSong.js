import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import { useState } from 'react'
import AudioButton from '../Buttons/AudioButton'
import ButtonSecondary from '../Buttons/ButtonSecondary'
import { XIcon, ResetIcon, BinIcon } from '../Icons'

SavedSong.propTypes = {
  stopPlayingSong: PropTypes.func,
  artist: PropTypes.string.isRequired,
  songTitle: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  deleteSavedSong: PropTypes.func,
  toggleCurrentSongId: PropTypes.func,
  isSongPlaying: PropTypes.bool.isRequired,
  currentSongId: PropTypes.string,
}

export default function SavedSong({
  stopPlayingSong,
  artist,
  songTitle,
  id,
  deleteSavedSong,
  toggleCurrentSongId,
  isSongPlaying,
  currentSongId,
}) {
  const [toBeDeleted, setToBeDeleted] = useState(false)

  return (
    <>
      {toBeDeleted || (
        <ListItem>
          <ButtonWrapper>
            <ButtonSecondary
              id={id}
              onClick={() => prepareToDelete(id)}
              data-testid="prepare-delete-button"
            >
              <XIcon />
            </ButtonSecondary>
          </ButtonWrapper>
          <Artist>{artist}</Artist>
          <Song>{songTitle}</Song>
          <AudioButtonWrapper>
            <AudioButton
              onClick={() => toggleCurrentSongId(id)}
              isSongPlaying={isSongPlaying && currentSongId === id}
              currentSongId={currentSongId}
            />
          </AudioButtonWrapper>
        </ListItem>
      )}
      {toBeDeleted && (
        <ConfirmationWrapper>
          <p>Do you want to delete this song?</p>
          <ConfirmationButton
            id={id}
            onClick={() => deleteSavedSong(id)}
            data-testid="delete-button"
          >
            <BinIcon />
          </ConfirmationButton>
          <ConfirmationButton
            id={id}
            onClick={() => setToBeDeleted(false)}
            data-testid="reset-button"
          >
            <ResetIcon />
          </ConfirmationButton>
        </ConfirmationWrapper>
      )}
    </>
  )

  function prepareToDelete(id) {
    stopPlayingSong(id)
    setToBeDeleted(true)
  }
}

const ListItem = styled.li`
  background-color: var(--primary-light);
  display: grid;
  column-gap: 14px;
  grid-template-columns: 36px auto 46px;
  grid-template-rows: 1fr 1fr;
  padding: 20px;
`
const ButtonWrapper = styled.div`
  grid-row: 1/3;
  grid-column-start: 1;
`
const Artist = styled.span`
  display: block;
  font-size: 75%;
  font-weight: 300;
  padding-bottom: 6px;
  text-transform: uppercase;
  grid-column-start: 2;
`
const Song = styled.span`
  color: var(--primary-dark);
  font-size: 1em;
  font-weight: 600;
  grid-column-start: 2;
`
const AudioButtonWrapper = styled.div`
  grid-column-start: 3;
  grid-row: 1/3;
  justify-self: end;
`
const ConfirmationWrapper = styled.section`
  background-color: var(--primary-light);
  display: grid;
  column-gap: 14px;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  padding: 20px;
  p {
    align-self: start;
    grid-column: 1/3;
    margin: 0;
    text-align: center;
  }
`
const ConfirmationButton = styled(ButtonSecondary)`
  justify-self: center;
`
