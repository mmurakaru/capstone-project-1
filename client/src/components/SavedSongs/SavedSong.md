One saved song item:

```js
<SavedSong
  stopPlayingSongById={() => {}}
  artist="Unprocessed"
  songTitle="Real"
  id="a"
  deleteSavedSong={() => {}}
  toggleCurrentSongId={() => {}}
  isSongPlaying={false}
  currentSongId="a"
/>
```
