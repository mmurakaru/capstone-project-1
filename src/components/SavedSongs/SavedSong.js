import styled from 'styled-components/macro'
import PropTypes from 'prop-types'
import { useState } from 'react'
import AudioButton from '../Buttons/AudioButton'
import Button from '../Buttons/Button'
import { XIcon, ResetIcon, BinIcon } from '../Icons'

SavedSong.propTypes = {
  stopPlayingSong: PropTypes.func,
  artist: PropTypes.string.isRequired,
  songTitle: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  deleteSavedSong: PropTypes.func,
  toggleCurrentSongId: PropTypes.func,
  isSongPlaying: PropTypes.bool.isRequired,
  currentSongId: PropTypes.string,
}

export default function SavedSong({
  stopPlayingSong,
  artist,
  songTitle,
  id,
  deleteSavedSong,
  toggleCurrentSongId,
  isSongPlaying,
  currentSongId,
}) {
  const [toBeDeleted, setToBeDeleted] = useState(false)

  return (
    <>
      {toBeDeleted || (
        <ListItem>
          <ButtonWrapper>
            <PrepareToDeleteButton
              isDelteButton
              id={id}
              onClick={() => prepareToDelete(id)}
              data-testid="prepare-delete-button"
            >
              <XIcon />
            </PrepareToDeleteButton>
          </ButtonWrapper>
          <Content>
            <Artist>{artist}</Artist>
            <Song>{songTitle}</Song>
            <AudioButtonWrapper>
              <AudioButton
                isSmall
                onClick={() => toggleCurrentSongId(id)}
                isSongPlaying={isSongPlaying && currentSongId === id}
                currentSongId={currentSongId}
              />
            </AudioButtonWrapper>
          </Content>
        </ListItem>
      )}
      {toBeDeleted && (
        <Container>
          <ConfirmationWrapper>
            <p>Do you want to delete {songTitle}?</p>
            <ConfirmationButtonWrapper>
              <DelteButton
                id={id}
                onClick={() => deleteSavedSong(id)}
                data-testid="delete-button"
              >
                <BinIcon />
              </DelteButton>
              <ResetButton
                id={id}
                onClick={() => setToBeDeleted(false)}
                data-testid="reset-button"
              >
                <ResetIcon />
              </ResetButton>
            </ConfirmationButtonWrapper>
          </ConfirmationWrapper>
        </Container>
      )}
    </>
  )

  function prepareToDelete(id) {
    stopPlayingSong(id)
    setToBeDeleted(true)
  }
}

const ListItem = styled.li`
  display: grid;
  column-gap: 8px;
  grid-template-columns: 48px auto;
  padding: 0 22px;
`
const ButtonWrapper = styled.div`
  grid-row: 1/3;
  grid-column-start: 1;
`
const PrepareToDeleteButton = styled(Button)`
  height: 48px;
  width: 48px;
`
const Content = styled.div`
  box-shadow: var(--shadow-light);
  border-radius: 10px 34px 34px 10px;
  display: grid;
  grid-template-columns: 82% auto;
  grid-template-rows: 50% 50%;
  padding: 0 0 0 8px;
`
const Artist = styled.span`
  align-self: end;
  color: var(--primary-regular);
  font-size: 0.9em;
  font-weight: 300;
  text-transform: uppercase;
`
const Song = styled.span`
  align-self: start;
  color: var(--primary-regular);
  grid-row-start: 2;
  font-size: 0.9em;
  font-weight: 700;
  text-transform: uppercase;
`
const AudioButtonWrapper = styled.div`
  grid-column-start: 3;
  grid-row: 1/3;
  justify-self: end;
`
const Container = styled.div`
  padding: 0 22px;
  display: grid;
`
const ConfirmationWrapper = styled.div`
  p {
    align-self: start;
    color: var(--primary-regular);
    font-size: 0.9em;
    font-weight: 700;
    text-align: center;
  }
`
const ConfirmationButtonWrapper = styled.div`
  display: flex;
  justify-content: space-evenly;
`
const DelteButton = styled(Button)`
  height: 48px;
  width: 48px;
  fill: var(--cta-red);
`
const ResetButton = styled(Button)`
  height: 48px;
  width: 48px;
  fill: var(--spotify-green);
`
