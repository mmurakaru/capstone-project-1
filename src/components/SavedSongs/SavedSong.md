One saved song item:

```js
<SavedSong
  artist="Unprocessed"
  songTitle="Real"
  id="a"
  deleteSavedSong={() => {}}
  toggleCurrentSongId={() => {}}
  isSongPlaying={false}
  currentSongId="a"
/>
```
