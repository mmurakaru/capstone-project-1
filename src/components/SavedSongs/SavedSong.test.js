import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import SavedSong from './SavedSong'

describe('SavedSong', () => {
  it('renders correctly', () => {
    const { container } = render(
      <SavedSong
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  it('calls toggleCurrentSongId with the correct id', () => {
    const toggleCurrentSongIdMock = jest.fn()
    const { getByTestId } = render(
      <SavedSong
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => toggleCurrentSongIdMock('a')}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    const button = getByTestId('audio-button')
    user.click(button)
    expect(toggleCurrentSongIdMock).toHaveBeenCalledWith('a')
  })
  it('has delete button', () => {
    const { getByTestId } = render(
      <SavedSong
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    const Button = getByTestId('x-button')
    expect(Button).toBeInTheDocument()
  })
  it('renders delete field', () => {
    const { getByTestId, getByText } = render(
      <SavedSong
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    const Button = getByTestId('x-button')
    user.click(Button)
    expect(getByText('Lorem ipsum')).toBeInTheDocument()
    expect(getByTestId('reset-button')).toBeInTheDocument()
    expect(getByTestId('delete-button')).toBeInTheDocument()
  })
  it('toggles the isSongPlaying state correctly', () => {
    const { getByTitle, queryByTitle, rerender } = render(
      <SavedSong
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={false}
        currentSongId="a"
      />
    )
    expect(getByTitle('play')).toBeInTheDocument()

    rerender(
      <SavedSong
        artist="Unprocessed"
        songTitle="Real"
        id="a"
        deleteSavedSong={() => {}}
        toggleCurrentSongId={() => {}}
        isSongPlaying={true}
        currentSongId="a"
      />
    )
    expect(queryByTitle('play')).not.toBeInTheDocument()
    expect(getByTitle('pause')).toBeInTheDocument()
  })
})
