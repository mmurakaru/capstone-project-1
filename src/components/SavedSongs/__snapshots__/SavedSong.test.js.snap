// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavedSong renders correctly 1`] = `
.c6 {
  background-color: transparent;
  border: var(--primary-dark) 2px solid;
  fill: var(--primary-dark);
  height: 46px;
  padding: 10px;
  width: 46px;
}

.c2 {
  background-color: var(--primary-dark);
  border: none;
  fill: var(--white);
  font-size: 0.8em;
  padding: 10px;
  height: 100%;
  width: 36px;
}

.c0 {
  background-color: var(--primary-light);
  display: grid;
  -webkit-column-gap: 14px;
  column-gap: 14px;
  grid-template-columns: 36px auto 46px;
  grid-template-rows: 1fr 1fr;
  padding: 20px;
}

.c1 {
  grid-row: 1/3;
  grid-column-start: 1;
}

.c3 {
  display: block;
  font-size: 75%;
  font-weight: 300;
  padding-bottom: 6px;
  text-transform: uppercase;
  grid-column-start: 2;
}

.c4 {
  color: var(--primary-dark);
  font-size: 1em;
  font-weight: 600;
  grid-column-start: 2;
}

.c5 {
  grid-column-start: 3;
  grid-row: 1/3;
  justify-self: end;
}

<li
  class="c0"
>
  <div
    class="c1"
  >
    <button
      class="c2"
      data-testid="x-button"
      id="a"
    >
      <svg>
        x.svg
      </svg>
    </button>
  </div>
  <span
    class="c3"
  >
    Unprocessed
  </span>
  <span
    class="c4"
  >
    Real
  </span>
  <div
    class="c5"
  >
    <button
      class="c6"
      data-testid="audio-button"
    >
      <svg
        title="play"
      >
        play.svg
      </svg>
    </button>
  </div>
</li>
`;
